File: mediator-pattern/mediator_pattern.rb
# What is the Mediator Design Pattern?
#
# - It is used to handle communication between related objects
# - All communication is handled by the Mediator and the Colleagues
#   don't need to know anything about each other
# - GOF: Allows loose coupling by encapsulating the way disparate
#   sets of objects interact and communicate with each other.
#   Allows for the actions of each object set to vary independently
#   of one another.
#
# @see http://www.newthinktank.com/2012/10/mediator-design-pattern-tutorial/

class StockOffer
  attr_reader :shares, :symbol, :colleague_code

  def initialize(shares, symbol, colleague_code)
    @shares = shares
    @symbol = symbol
    @colleague_code = colleague_code
  end
end

class Colleague
  attr_accessor :mediator, :code

  def initialize(mediator)
    @mediator = mediator
    @mediator.add_colleague self
  end

  def sale_offer(stock, shares)
    mediator.sale_offer(stock, shares, code)
  end

  def buy_offer(stock, shares)
    mediator.buy_offer(stock, shares, code)
  end
end

class GormanSlacks < Colleague
  def initialize(mediator)
    super
    puts 'Gorman Slacks singed up for the exchange'
  end
end

class JTPoorman < Colleague
  def initialize(mediator)
    super
    puts 'JT Poorman singed up for the exchange'
  end
end

class StockMediator
  attr_reader :colleagues, :buy_offers, :sale_offers, :code

  def initialize
    @code = 0
    @colleagues = []
    @buy_offers = []
    @sale_offers = []
  end

  def add_colleague(colleague)
    @colleagues << colleague
    colleague.code = @code
    @code += 1
  end

  def sale_offer(stock, shares, code)
    offer = StockOffer.new(shares, stock, code)
    @sale_offers << offer

    @buy_offers.each do |buy_offer|
      if offer_match? buy_offer, offer
        buy(buy_offer(offer))
        break
      end
    end
  end

  def buy_offer(stock, shares, code)
    offer = StockOffer.new(shares, stock, code)
    @buy_offers << offer

    @sale_offers.each do |sale_offer|
      buy(offer, sale_offer) if offer_match? sale_offer, offer
    end
  end

  private

  def offer_match?(offer1, offer2)
    offer1.symbol == offer2.symbol && offer1.shares == offer2.shares && offer1.colleague_code != offer2.colleague_code
  end

  def matching_offer(offer)
    @sale_offers.select { |o| offer_match? o, offer }.first
  end

  def buy(buy_offer, sale_offer)
    @sale_offers.delete sale_offer
    @buy_offers.delete buy_offer

    puts "#{sale_offer.shares} shares of #{sale_offer.symbol} sold to colleague of code #{buy_offer.colleague_code}"
  end
end

################################################################################

nyse = StockMediator.new
broker = GormanSlacks.new(nyse)
broker2 = JTPoorman.new(nyse)

broker.sale_offer('MSFT', 100)
broker.sale_offer('GOOG', 50)

broker2.sale_offer('NRG', 10)

broker.buy_offer('NRG', 10)
broker2.buy_offer('MSFT', 100)

puts "\nsale offers " + '=' * 15
puts nyse.sale_offers.map { |o| "#{o.symbol} - #{o.shares}" }
puts "\nbuy offers  " + '=' * 15
puts nyse.buy_offers.map { |o| "#{o.symbol} - #{o.shares}" }

File: mediator-pattern/mediator-pattern.ts
/**
 * What is the Mediator Design Pattern?
 *
 * - It is used to handle communication between related objects
 * - All communication is handled by the Mediator and the Colleagues
 *   don't need to know anything about each other
 * - GOF: Allows loose coupling by encapsulating the way disparate
 *   sets of objects interact and communicate with each other.
 *   Allows for the actions of each object set to vary independently
 *   of one another.
 *
 * @see http://www.newthinktank.com/2012/10/mediator-design-pattern-tutorial/
 */

class StockOffer {
  public constructor(
    readonly stockShares: number,
    readonly stockSymbol: string,
    readonly colleagueCode: number
  ) {}
}

abstract class Colleague {
  private colleagueCode: number = -1;

  public constructor(private readonly mediator: Mediator) {
    mediator.addColleague(this);
  }

  public saleOffer(stock: string, shares: number): void {
    this.mediator.saleOffer(stock, shares, this.colleagueCode);
  }

  public buyOffer(stock: string, shares: number): void {
    this.mediator.buyOffer(stock, shares, this.colleagueCode);
  }

  public setColleagueCode(colleagueCode: number): void {
    this.colleagueCode = colleagueCode;
  }
}

class GormanSlacks extends Colleague {
  public constructor(mediator: Mediator) {
    super(mediator);
    console.log("Gorman Slacks signed up for the exchange");
  }
}

class JTPoorman extends Colleague {
  public constructor(mediator: Mediator) {
    super(mediator);
    console.log("JTPoormanSlacks Slacks signed up for the exchange");
  }
}

interface Mediator {
  saleOffer(stock: string, shares: number, colleagueCode: number): void;
  buyOffer(stock: string, shares: number, colleagueCode: number): void;
  addColleague(colleague: Colleague): void;
}

class StockMediator implements Mediator {
  private colleagueCode: number = 0;
  private colleagues: Record<number, Colleague> = {};
  private stockBuyOffers: StockOffer[] = [];
  private stockSaleOffers: StockOffer[] = [];

  saleOffer(stock: string, shares: number, colleagueCode: number): void {
    let stockSold: boolean = false;

    for (let offer of this.stockBuyOffers) {
      if (offer.stockSymbol === "stock" && offer.stockShares === shares) {
        console.log(
          `${shares} shares of ${stock} sold to colleague code ${colleagueCode}`
        );
        this.stockBuyOffers.splice(this.stockBuyOffers.indexOf(offer), 1);
        stockSold = true;
        break;
      }
    }

    if (!stockSold) {
      console.log(`${shares} shares of ${stock} added to inventory`);
      const newOffer: StockOffer = new StockOffer(shares, stock, colleagueCode);
      this.stockSaleOffers.push(newOffer);
    }
  }

  buyOffer(stock: string, shares: number, colleagueCode: number): void {
    let stockBought: boolean = false;

    for (let offer of this.stockSaleOffers) {
      if (offer.stockSymbol === stock && offer.stockShares === shares) {
        console.log(
          `${shares} shares of ${stock} bought by colleague code ${colleagueCode}`
        );

        this.stockSaleOffers.splice(this.stockSaleOffers.indexOf(offer), 1);
        stockBought = true;
        break;
      }
    }

    if (!stockBought) {
      console.log(`${shares} shares of ${stock} added to inventory`);
      const newOffer: StockOffer = new StockOffer(shares, stock, colleagueCode);
      this.stockBuyOffers.push(newOffer);
    }
  }

  addColleague(colleague: Colleague): void {
    this.colleagues[this.colleagueCode] = colleague;
    colleague.setColleagueCode(this.colleagueCode);
    this.colleagueCode++;
  }

  getStockOfferings(): void {
    console.log("Stocks for sale");
    for (let offer of this.stockSaleOffers) {
      console.log(`\t- ${offer.stockShares} of ${offer.stockSymbol}`);
    }

    console.log("Stocks buy offers:");
    for (let offer of this.stockBuyOffers) {
      console.log(`\t- ${offer.stockShares} of ${offer.stockSymbol}`);
    }
  }
}

//------------------------------------------------------------------------------

const nyse: StockMediator = new StockMediator();
const broker: GormanSlacks = new GormanSlacks(nyse);
const broker2: JTPoorman = new JTPoorman(nyse);

broker.saleOffer("MSFT", 100);
broker.saleOffer("GOOG", 50);
broker2.buyOffer("MSFT", 100);
broker2.saleOffer("NRG", 10);
broker.buyOffer("NRG", 10);
nyse.getStockOfferings();
